File: Magestorm2/Server/MagestormServer/src/ProfanityChecker.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Dictionary;
import java.util.HashSet;
import java.util.Locale;
import java.util.Scanner;

public class ProfanityChecker {
    private static HashSet<String> _prohibitedTerms;

    public static void Init(String termsPath){
        _prohibitedTerms = new HashSet<>();
        File termsFile = new File(termsPath);
        try {
            Scanner termsScanner = new Scanner(termsFile);
            while(termsScanner.hasNextLine()){
                _prohibitedTerms.add(termsScanner.nextLine().toLowerCase(Locale.ROOT));
            }
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
        LetterReplace("a", "4");
        LetterReplace("b", "8");
        LetterReplace("e", "3");
        LetterReplace("g", "6");
        LetterReplace("i", "1");
        LetterReplace("l", "1");
        LetterReplace("o", "0");
        Main.LogMessage("Profanity list count: " + _prohibitedTerms.size());
    }
    private static void LetterReplace(String letter, String number){
        HashSet<String> additionalTerms = new HashSet<>();
        for (String prohibited : _prohibitedTerms){
            additionalTerms.add(prohibited.replace(letter, number));
        }
        for (String prohibited : additionalTerms){
            _prohibitedTerms.add(prohibited);
        }
    }
    public static boolean ContainsProhibitedLanguage(String text){
        for (String prohibited : _prohibitedTerms){
            if(text.contains(prohibited)){
                return true;
            }
        }
        return false;
    }
}

File: Magestorm2/Assets/Utility/ProfanityChecker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using UnityEngine;
public static class ProfanityChecker
{
    private static HashSet<string> _prohibitedTerms;
    public static void Init()
    {
        _prohibitedTerms = new HashSet<string>();
        string path = Application.dataPath + "/prohibitedterms.txt";
        string[] terms = File.ReadAllLines(path);
        foreach (string term in terms)
        {
            _prohibitedTerms.Add(term);
        }
        LetterReplace("a", "4");
        LetterReplace("b", "8");
        LetterReplace("e", "3");
        LetterReplace("g", "6");
        LetterReplace("i", "1");
        LetterReplace("l", "1");
        LetterReplace("o", "0");
    }

    private static void LetterReplace(string letter, string number)
    {
        HashSet<string> additionalTerms = new HashSet<string>();
        foreach(string term in _prohibitedTerms)
        {
            additionalTerms.Add(term.Replace(letter, number));
        }
        foreach(string term in additionalTerms)
        {
            if (!_prohibitedTerms.Contains(term))
            {
                _prohibitedTerms.Add(term);
            }
        }
    }
    public static bool ContainsProhibitedLanguage(string text)
    {
        string lower = text.ToLower();
        foreach(string term in _prohibitedTerms)
        {
            if (lower.Contains(term))
            {
                return true;
            }
        }
        return false;
    }
}

